homeassistant/components/reolink/__init__.py

```diff
@@ -0,0 +1,108 @@
+"""Reolink integration for HomeAssistant."""
+
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import cast
+
+import async_timeout
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    DEVICE_CONFIG_UPDATE_COORDINATOR,
+    DEVICE_UPDATE_INTERVAL,
+    DOMAIN,
+    HOST,
+    PLATFORMS,
+    SERVICE_PTZ_CONTROL,
+    SERVICE_SET_BACKLIGHT,
+    SERVICE_SET_DAYNIGHT,
+    SERVICE_SET_SENSITIVITY,
+)
+from .host import ReolinkHost
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Reolink from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    host = ReolinkHost(hass, cast(dict, entry.data), cast(dict, entry.options))
+
+    try:
+        if not await host.init():
+            raise ConfigEntryNotReady(
+                f"Error while trying to setup {host.api.host}:{host.api.port}: failed to obtain data from device."
+            )
+    except Exception as error:  # pylint: disable=broad-except
+        err = str(error)
+        if err:
+            raise ConfigEntryNotReady(
+                f'Error while trying to setup {host.api.host}:{host.api.port}: "{err}".'
+            ) from error
+        raise ConfigEntryNotReady(
+            f"Error while trying to setup {host.api.host}:{host.api.port}: failed to connect to device."
+        ) from error
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, host.stop)
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {HOST: host}
```

Comment:
Ideally we store a `dataclass` instead of a `dict` since it makes it easier to type.

Example [in `lookin/models.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/lookin/models.py)

Reply:
Thanks for the hint, I have updated the PR to use such a dataclass